File: testing/unit-tests/pkmntest/csharp/PokemonIOTests.cs
/*
 * Copyright (c) 2017-2018 Nicholas Corgan (n.corgan@gmail.com)
 *
 * Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
 * or copy at http://opensource.org/licenses/MIT)
 */

using System;
using NUnit.Framework;

namespace PKMNTest
{

public class PokemonIOTests
{
    private static System.Random rng = new System.Random();

    private static string LibPKMNTestFiles = Environment.GetEnvironmentVariable("LIBPKMN_TEST_FILES");
    private static string TmpDir = PKMN.Paths.GetTmpDir();

    private static void TestLoadingAndSaving(
        PKMN.Game game,
        string extension
    )
    {
        string tmpPath = System.IO.Path.Combine(
                             TmpDir,
                             String.Format("{0}_{1}.{2}",
                                 game,
                                 rng.Next(),
                                 extension
                             )
                          );

        PKMN.ItemEnumList itemList = PKMN.Database.Lists.ItemList(game);
        PKMN.MoveEnumList moveList = PKMN.Database.Lists.MoveList(game);
        PKMN.SpeciesEnumList pokemonList = PKMN.Database.Lists.PokemonList(
                                               Util.GameToGeneration(game),
                                               true
                                           );

        PKMN.Pokemon randomPokemon = Util.GetRandomPokemon(game, itemList, moveList, pokemonList);
        randomPokemon.ExportToFile(tmpPath);

        PKMN.Pokemon importedPokemon = new PKMN.Pokemon(tmpPath);
        Util.ComparePokemon(randomPokemon, importedPokemon);

        System.IO.File.Delete(tmpPath);
    }

    // Generation I

    public static void TestLoadingAndSavingPK1(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "pk1");
    }

    // Generation III

    public static void TestLoadingAndSavingPK2(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "pk2");
    }

    // Game Boy Advance

    public static void TestLoadingAndSaving3GPKM(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "3gpkm");
    }

    public static void TestLoadingAndSavingPK3(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "pk3");
    }

    // Gamecube

    public static void TestLoadingAndSavingCK3(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "ck3");
    }

    public static void TestLoadingAndSavingXK3(
        PKMN.Game game
    )
    {
        TestLoadingAndSaving(game, "xk3");
    }

    // Outside files

    public static void TestOutside3GPKM()
    {
        /*
         * Test files in repo and compare to known values.
         */
        string _3GPKMDir = System.IO.Path.Combine(LibPKMNTestFiles, "3gpkm");

        PKMN.Pokemon mightyena = new PKMN.Pokemon(
                                          System.IO.Path.Combine(_3GPKMDir, "MIGHTYENA.3gpkm")
                                      );
        Assert.AreEqual(mightyena.Species, PKMN.Species.MIGHTYENA);
        Assert.AreEqual(mightyena.Game, PKMN.Game.EMERALD);
        Assert.AreEqual(mightyena.Form, "Standard");
        Assert.AreEqual(mightyena.Nickname, "MIGHTYENA");
        Assert.IsFalse(mightyena.IsShiny);
        Assert.AreEqual(mightyena.HeldItem, PKMN.Item.HEART_SCALE);
        Assert.AreEqual(mightyena.Condition, PKMN.Condition.NONE);
        Assert.AreEqual(mightyena.OriginalTrainerName, "A");
        Assert.AreEqual(mightyena.OriginalTrainerPublicID, 61415);
        Assert.AreEqual(mightyena.OriginalTrainerSecretID, 3417);
        Assert.AreEqual(mightyena.OriginalTrainerID, 223997927);
        Assert.AreEqual(mightyena.OriginalTrainerGender, PKMN.Gender.FEMALE);
        Assert.AreEqual(mightyena.CurrentTrainerFriendship, 254);
        Assert.AreEqual(mightyena.Ability, PKMN.Ability.INTIMIDATE);
        Assert.AreEqual(mightyena.Ball, PKMN.Ball.GREAT_BALL);
        Assert.AreEqual(mightyena.LevelMet, 25);
        Assert.AreEqual(mightyena.LocationMet, "Route 120");
        Assert.AreEqual(mightyena.OriginalGame, PKMN.Game.EMERALD);
        Assert.AreEqual(mightyena.Personality, 3557601241);
        Assert.AreEqual(mightyena.Experience, 128734);
        Assert.AreEqual(mightyena.Level, 50);

        Assert.AreEqual(mightyena.Markings.Count, 4);
        foreach(PKMN.Marking marking in mightyena.Markings.Keys)
        {
            Assert.IsFalse(mightyena.Markings[marking]);
        }

        Assert.AreEqual(mightyena.Ribbons.Count, 32);
        foreach(string ribbon in mightyena.Ribbons.Keys)
        {
            if(ribbon.Equals("Champion")) {
                Assert.IsTrue(mightyena.Ribbons[ribbon]);
            } else {
                Assert.IsFalse(mightyena.Ribbons[ribbon]);
            }
        }

        Assert.AreEqual(mightyena.ContestStats.Count, 6);
        foreach(PKMN.ContestStat contestStat in mightyena.ContestStats.Keys)
        {
            Assert.AreEqual(mightyena.ContestStats[contestStat], 0);
        }

        PKMN.Move[] expectedMightyenaMoves =
        {
            PKMN.Move.CRUNCH,
            PKMN.Move.STRENGTH,
            PKMN.Move.SHADOW_BALL,
            PKMN.Move.DOUBLE_EDGE
        };
        Assert.AreEqual(mightyena.Moves.Count, 4);
        for(int i = 0; i < 4; ++i)
        {
            Assert.AreEqual(mightyena.Moves[i].Move, expectedMightyenaMoves[i]);
        }

        Assert.AreEqual(mightyena.EVs.Count, 6);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.HP], 30);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.ATTACK], 110);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.DEFENSE], 32);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.SPEED], 48);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.SPECIAL_ATTACK], 17);
        Assert.AreEqual(mightyena.EVs[PKMN.Stat.SPECIAL_DEFENSE], 83);

        Assert.AreEqual(mightyena.IVs.Count, 6);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.HP], 26);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.ATTACK], 28);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.DEFENSE], 4);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.SPEED], 13);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.SPECIAL_ATTACK], 25);
        Assert.AreEqual(mightyena.IVs[PKMN.Stat.SPECIAL_DEFENSE], 26);

        Assert.AreEqual(mightyena.Stats.Count, 6);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.HP], 146);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.ATTACK], 122);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.DEFENSE], 81);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.SPEED], 87);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.SPECIAL_ATTACK], 79);
        Assert.AreEqual(mightyena.Stats[PKMN.Stat.SPECIAL_DEFENSE], 88);
    }
}

}

File: testing/unit-tests/pkmntest/lua/pokemon_io_tests.lua
--
-- Copyright (c) 2017-2018 Nicholas Corgan (n.corgan@gmail.com)
--
-- Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
-- or copy at http://opensource.org/licenses/MIT)
--

local pkmn = require("pkmn")
local pkmntest_utils = require("pkmntest_utils")

local luaunit = require("luaunit")

local LIBPKMN_TEST_FILES = pkmn.paths.getenv("LIBPKMN_TEST_FILES")
local TMP_DIR = pkmn.paths.get_tmp_dir()

local pokemon_io_tests = {}

function pokemon_io_tests.test_saving_and_loading(game, extension)
    local tmp_path = pkmntest_utils.concat_path(
                         TMP_DIR,
                         string.format(
                             "%s_%d.%s",
                             game,
                             math.random(0, 0xFFFF),
                             extension
                         )
                     )

    local item_list = pkmn.database.get_item_list(game)
    local move_list = pkmn.database.get_move_list(game)
    local pokemon_list = pkmn.database.get_pokemon_list(
                             pkmntest_utils.GAME_TO_GENERATION[game],
                             true
                         )

    local random_pokemon = pkmntest_utils.get_random_pokemon(game, pokemon_list, move_list, item_list)
    random_pokemon:export_to_file(tmp_path)

    local imported_pokemon = pkmn.pokemon(tmp_path)
    pkmntest_utils.compare_pokemon(imported_pokemon, random_pokemon)

    os.remove(tmp_path)
end

-- Generation I

function test_red_saving_and_loading_pk1()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.RED, "pk1")
end

function test_blue_saving_and_loading_pk1()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.BLUE, "pk1")
end

function test_yellow_saving_and_loading_pk1()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.YELLOW, "pk1")
end

-- Generation II

function test_gold_saving_and_loading_pk2()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.GOLD, "pk2")
end

function test_silver_saving_and_loading_pk2()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.SILVER, "pk2")
end

function test_crystal_saving_and_loading_pk2()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.CRYSTAL, "pk2")
end

-- Game Boy Advance

function test_ruby_saving_and_loading_3gpkm()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.RUBY, "3gpkm")
end

function test_ruby_saving_and_loading_pk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.RUBY, "pk3")
end

function test_sapphire_saving_and_loading_3gpkm()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.SAPPHIRE, "3gpkm")
end

function test_sapphire_saving_and_loading_pk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.SAPPHIRE, "pk3")
end

function test_emerald_saving_and_loading_3gpkm()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.EMERALD, "3gpkm")
end

function test_emerald_saving_and_loading_pk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.EMERALD, "pk3")
end

function test_firered_saving_and_loading_3gpkm()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.FIRERED, "3gpkm")
end

function test_firered_saving_and_loading_pk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.FIRERED, "pk3")
end

function test_leafgreen_saving_and_loading_3gpkm()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.LEAFGREEN, "3gpkm")
end

function test_leafgreen_saving_and_loading_pk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.LEAFGREEN, "pk3")
end

-- Gamecube

function test_colosseum_saving_and_loading_ck3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.COLOSSEUM, "ck3")
end

function test_xd_saving_and_loading_xk3()
    pokemon_io_tests.test_saving_and_loading(pkmn.game.XD, "xk3")
end

-- Outside files

function test_outside_3gpkm()
    local CONTEST_LEVELS = {"Super", "Hyper", "Master"}
    local RIBBONS = {"Champion", "Winning", "Victory", "Artist",
                     "Effort", "Marine", "Land", "Sky",
                     "Country", "National", "Earth", "World"}
    --
    -- Test files in repo and compare to known values.
    --
    local _3gpkm_dir = pkmntest_utils.concat_path(LIBPKMN_TEST_FILES, "3gpkm")

    local mightyena = pkmn.pokemon(pkmntest_utils.concat_path(_3gpkm_dir, "MIGHTYENA.3gpkm"))
    luaunit.assertEquals(mightyena.species, pkmn.species.MIGHTYENA)
    luaunit.assertEquals(mightyena.game, pkmn.game.EMERALD)
    luaunit.assertEquals(mightyena.form, "Standard")
    luaunit.assertEquals(mightyena.nickname, "MIGHTYENA")
    luaunit.assertFalse(mightyena.is_shiny)
    luaunit.assertEquals(mightyena.condition, pkmn.condition.NONE)
    luaunit.assertEquals(mightyena.held_item, pkmn.item.HEART_SCALE)
    luaunit.assertEquals(mightyena.original_trainer_name, "A")
    luaunit.assertEquals(mightyena.original_trainer_public_id, 61415)
    luaunit.assertEquals(mightyena.original_trainer_secret_id, 3417)
    luaunit.assertEquals(mightyena.original_trainer_id, 223997927)
    luaunit.assertEquals(mightyena.original_trainer_gender, pkmn.gender.FEMALE)
    luaunit.assertEquals(mightyena.current_trainer_friendship, 254)
    luaunit.assertEquals(mightyena.ability, pkmn.ability.INTIMIDATE)
    luaunit.assertEquals(mightyena.ball, pkmn.item.GREAT_BALL)
    luaunit.assertEquals(mightyena.level_met, 25)
    luaunit.assertEquals(mightyena.location_met, "Route 120")
    luaunit.assertEquals(mightyena.original_game, pkmn.game.EMERALD)
    luaunit.assertEquals(mightyena.personality, 3557601241)
    luaunit.assertEquals(mightyena.experience, 128734)
    luaunit.assertEquals(mightyena.level, 50)

    luaunit.assertEquals(#mightyena.markings, 4)
    for i = 1, #mightyena.markings.keys
    do
        luaunit.assertFalse(mightyena.markings[mightyena.markings.keys[i]])
    end

    luaunit.assertEquals(#mightyena.ribbons, 32)
    -- Contest ribbons
    for contest_type_index = 1, #mightyena.contest_stats.keys
    do
        if mightyena.contest_stats.keys[contest_type_index] ~= pkmn.contest_stat.FEEL
        then
            local contest_stat_name = pkmn.contest_stat_to_string(mightyena.contest_stats.keys[contest_type_index])

            luaunit.assertFalse(mightyena.ribbons[contest_stat_name])
            for contest_level_index = 1, #CONTEST_LEVELS
            do
                local ribbon_name = contest_stat_name .. " " .. CONTEST_LEVELS[contest_level_index]
                luaunit.assertFalse(mightyena.ribbons[ribbon_name])
            end
        end
    end
    -- Other ribbons
    for ribbon_index = 1, #RIBBONS
    do
        if RIBBONS[ribbon_index] == "Champion"
        then
            luaunit.assertTrue(mightyena.ribbons[RIBBONS[ribbon_index]])
        else
            luaunit.assertFalse(mightyena.ribbons[RIBBONS[ribbon_index]])
        end
    end

    luaunit.assertEquals(#mightyena.contest_stats, 6)
    for contest_type_index = 1, #mightyena.contest_stats.keys
    do
        luaunit.assertEquals(mightyena.contest_stats[mightyena.contest_stats.keys[contest_type_index]], 0)
    end

    local expected_mightyena_moves =
    {
        pkmn.move.CRUNCH,
        pkmn.move.STRENGTH,
        pkmn.move.SHADOW_BALL,
        pkmn.move.DOUBLE_EDGE
    }
    luaunit.assertEquals(#mightyena.moves, 4)
    for move_index = 1, 4
    do
        luaunit.assertEquals(mightyena.moves[move_index].move, expected_mightyena_moves[move_index])
    end

    luaunit.assertEquals(#mightyena.EVs, 6)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.HP], 30)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.ATTACK], 110)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.DEFENSE], 32)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.SPEED], 48)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.SPECIAL_ATTACK], 17)
    luaunit.assertEquals(mightyena.EVs[pkmn.stat.SPECIAL_DEFENSE], 83)

    luaunit.assertEquals(#mightyena.IVs, 6)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.HP], 26)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.ATTACK], 28)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.DEFENSE], 4)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.SPEED], 13)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.SPECIAL_ATTACK], 25)
    luaunit.assertEquals(mightyena.IVs[pkmn.stat.SPECIAL_DEFENSE], 26)

    luaunit.assertEquals(#mightyena.stats, 6)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.HP], 146)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.ATTACK], 122)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.DEFENSE], 81)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.SPEED], 87)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.SPECIAL_ATTACK], 79)
    luaunit.assertEquals(mightyena.stats[pkmn.stat.SPECIAL_DEFENSE], 88)
end
